### PR TITLE
feat: introduce prometheus and grafana as metrics service

### PR DESCRIPTION
**Pull request by ytaoeer**
Changes:

- Add download and start / kill scripts in cargo make file.
- Add prometheus service for wheel.

Ref: #66 .

### CODE REVIEW

**Comment by ytaoeer** [comment]
prometheus

**Comment by MrCroxx** [comment]
Thank you for your contribution. Please refine the PR title and message. Describe the motivation of this pull request, what changed in this PR and list references of related issues and PRs.

File: common/src/config.rs [hunk 1]
```diff
@@ -65,6 +65,13 @@ pub struct Node {
     pub port: u16,
 }
 
+// prometheus config
```
Comment (MrCroxx) [inline common/src/config.rs:68]: Remove the comment.

File: make/grafana.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+GRAFANA_SYSTEM = "${SYSTEM_AMD64}"
+GRAFANA_DOWNLOAD_PATH = "${PREFIX_TMP}/grafana.tar.gz"
+GRAFANA_VERSION = "8.3.3"
+GRAFANA_RELEASE = "grafana-${GRAFANA_VERSION}"
+GRAFANA_DOWNLOAD_TAR_GZ = "https://dl.grafana.com/oss/release/${GRAFANA_RELEASE}.${GRAFANA_SYSTEM}.tar.gz"
+
+[tasks.download-grafana]
+category = "Grafana"
+dependencies = ["prepare"]
+#condition = { env_set = [ "ENABLE_PROMETHEUS_GRAFANA" ] }
```
Comment (MrCroxx) [inline make/grafana.toml:13]: Remove the comment.

File: Makefile.toml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-extend = [{ path = "make/common.toml" }, { path = "make/minio.toml" }, { path = "make/jaeger.toml" }]
+extend = [{ path = "make/common.toml" }, { path = "make/minio.toml" }, { path = "make/jaeger.toml" },{ path = "make/prometheus.toml"}]
```
Comment (MrCroxx) [inline Makefile.toml:1]: Add `grafana.toml` here.

File: Makefile.toml [hunk 4]
```diff
@@ -38,7 +38,7 @@ echo success
 [tasks.pre-dev]
 category = "Develop"
 description = "Downloads necessary tools and build required components."
-dependencies = ["download-minio", "download-mcli", "clean-data"]
+dependencies = ["download-minio", "download-mcli","download-prometheus", "clean-data"]
```
Comment (MrCroxx) [inline Makefile.toml:41]: Ditto.

File: make/grafana.toml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+set -e
+if [ -d "${PREFIX_BIN}/grafana" ]; then
+    exit 0
+fi
+echo "Grafana not found, download ${GRAFANA_RELEASE}"
+
+curl -fL -o "${GRAFANA_DOWNLOAD_PATH}" "${GRAFANA_DOWNLOAD_TAR_GZ}"
+tar -xf "${GRAFANA_DOWNLOAD_PATH}" -C "${PREFIX_TMP}"
+mv "${PREFIX_TMP}/${GRAFANA_RELEASE}" "${PREFIX_BIN}/grafana"
+#rm "${GRAFANA_DOWNLOAD_PATH}"
```
Comment (MrCroxx) [inline make/grafana.toml:27]: Why comment this line? If you don't need it please just remove it.

File: make/prometheus.toml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+fi
+echo ${PROMETHEUS_DOWNLOAD_PATH}
+tar -xf "${PROMETHEUS_DOWNLOAD_PATH}" -C "${PREFIX_TMP}"
+echo ${PREFIX_TMP}
+mv "${PREFIX_TMP}/${PROMETHEUS_RELEASE}" "${PREFIX_BIN}/prometheus"
+cp /home/ivic/workspace/RunKV/make/prometheus.yml  "${PREFIX_BIN}/prometheus/"
+cd "${PREFIX_BIN}/prometheus"
+./prometheus --config.file=prometheus.yml --web.listen-address=:9091
+cd "${CUR_DIR}"
+#rm ${PROMETHEUS_DOWNLOAD_PATH}
```
Comment (MrCroxx) [inline make/prometheus.toml:35]: Ditto.

File: make/prometheus.yml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+# my global config
```
Comment (MrCroxx) [inline make/prometheus.yml:1]: Configuration files should not be kept in `make` directory. `make` directory is used to keep make scripts. Please move this file to `etc`.

Besides, remove the comment.

Comment (MrCroxx) [reply make/prometheus.yml:1]: Besides, remove all **unused** comments and fields in this file.

Comment (MrCroxx) [reply make/prometheus.yml:1]: BTW, `tests/etc` also need one copy of this file.

File: make/prometheus.yml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+  #   # metrics_path defaults to '/metrics'
+  #   # scheme defaults to 'http'.
+
+  #   static_configs:
+  #     - targets: ["192.168.105.132:9091"]
+  - job_name: "test_node"
+    scrape_interval: 0.01s
+    static_configs:
+      - targets: ["192.168.105.132:9898"]
```
Comment (MrCroxx) [inline make/prometheus.yml:33]: Please don't include ip address of your development environment. If this file is a configuration template, use localhost or "127.0.0.1" instead.

File: tests/etc/wheel.toml [hunk 9]
```diff
@@ -28,3 +28,7 @@ meta_cache_capacity = "1 KiB"
 log_dir_path = "/path/to/log/dir"
 log_file_capacity = "64 MiB"
 block_cache_capacity = "256 MiB"
+
+[prometheus]
+host = "192.168.105.132"
```
Comment (MrCroxx) [inline tests/etc/wheel.toml:33]: Ditto.

File: wheel/Cargo.toml [hunk 10]
```diff
@@ -37,6 +37,13 @@ tonic = "0.6.2"
 tracing = "0.1"
 tracing-subscriber = "0.3"
 
+prometheus = "0.13.0"
+lazy_static = "^1.4"
+chrono = { version = "0.4", features = ["serde"] }
+hyper = { version = "^0.14", features = ["server", "http1", "tcp"] }
+http = "0.2.6"
+log = "0.4.16"
```
Comment (MrCroxx) [inline wheel/Cargo.toml:45]: We don't need `log` crate. In RunKV we use `tracing` to record event and span logs. Please remove the dependency.

File: wheel/src/config.rs [hunk 11]
```diff
@@ -1,4 +1,4 @@
-use runkv_common::config::{CacheConfig, LsmTreeConfig, MinioConfig, Node, S3Config};
+use runkv_common::config::{CacheConfig, LsmTreeConfig, MinioConfig, Node, S3Config,PrometheusConfig};
```
Comment (MrCroxx) [inline wheel/src/config.rs:1]: Please make sure you've lint the code.

File: wheel/src/lib.rs [hunk 12]
```diff
@@ -33,6 +34,38 @@ use worker::heartbeater::{Heartbeater, HeartbeaterOptions};
 
 use crate::config::WheelConfig;
 
+use chrono::{DateTime, Duration, Local, NaiveDateTime};
+use http;
+// prometheus
```
Comment (MrCroxx) [inline wheel/src/lib.rs:39]: Remove the comment.

File: wheel/src/lib.rs [hunk 13]
```diff
@@ -33,6 +34,38 @@ use worker::heartbeater::{Heartbeater, HeartbeaterOptions};
...
+        labels! {"handler" => "all",}
+    ))
+    .unwrap();
+    static ref EXAMPLE_TIME_GAUGE: Gauge = register_gauge!(opts!(
+        "example_time",
+        "The example time.",
+        labels! {"handler" => "all",}
+    ))
+    .unwrap();
+}
```
Comment (MrCroxx) [inline wheel/src/lib.rs:66]: Please refactor the field names with meaningful description.

File: wheel/src/lib.rs [hunk 14]
```diff
@@ -44,6 +77,10 @@ pub async fn bootstrap_wheel(
         tokio::spawn(async move { worker.run().await });
     }
 
+    test_prometheus();
+    let listen_addr: String = format!("{}:{}", config.prometheus.host, config.prometheus.port);
```
Comment (MrCroxx) [inline wheel/src/lib.rs:81]:
```suggestion
    let listen_addr = format!("{}:{}", config.prometheus.host, config.prometheus.port);
```

Please remove unnecessary type identifier.

File: wheel/src/lib.rs [hunk 15]
```diff
@@ -53,6 +90,38 @@ pub async fn bootstrap_wheel(
         .map_err(Error::err)
 }
 
+pub fn test_prometheus() {
+    tokio::spawn(async move {
+        log::info!("every second update example_num and exampe_time");
```
Comment (MrCroxx) [inline wheel/src/lib.rs:95]: Use tracing instead.

File: wheel/src/lib.rs [hunk 16]
```diff
@@ -53,6 +90,38 @@ pub async fn bootstrap_wheel(
...
+        loop {
+            EXAMPLE_COUNTER.inc();
+            let dt = Local::now();
+            let t = dt.timestamp_millis();
+            EXAMPLE_TIME_GAUGE.set(t as f64);
+            sleep(tokio::time::Duration::from_millis(10)).await;
+        }
+    });
+}
```
Comment (MrCroxx) [inline wheel/src/lib.rs:104]: Why we need test here?

1. Tests should not be placed in `lib.rs`, it's only used for building and bootstrapping nodes.
2. I think we don't need to unit test to test if prometheus works. Besides, this test cannot give you any useful feedbacks.
3. If you want to add a unit test, Please use `#[cfg(test)]` `#[test]` and wrap them in `tests` module to make it only available when testing. Please refer to the Rust book.

File: wheel/src/lib.rs [hunk 17]
```diff
@@ -53,6 +90,38 @@ pub async fn bootstrap_wheel(
...
+
+pub fn boot_prometheus_service(listen_addr: String, wheel: Wheel) {
+    tokio::spawn(async move {
+        log::info!(
```
Comment (MrCroxx) [inline wheel/src/lib.rs:108]: Ditto.

File: wheel/src/lib.rs [hunk 18]
```diff
@@ -53,6 +90,38 @@ pub async fn bootstrap_wheel(
...
+            listen_addr
+        );
+        let listen_prometheus_addr: SocketAddr = listen_addr.parse().unwrap();
+        let serve_future =
+            hyper::Server::bind(&listen_prometheus_addr).serve(make_service_fn(|_| async {
+                Ok::<_, hyper::Error>(service_fn(Wheel::prometheus_serve_req))
+            }));
+
+        if let Err(err) = serve_future.await {
+            eprintln!("server error: {}", err);
```
Comment (MrCroxx) [inline wheel/src/lib.rs:119]: Ditto.

File: wheel/src/service.rs [hunk 19]
```diff
@@ -34,6 +36,17 @@ use crate::components::Raft;
...
 use crate::meta::MetaStoreRef;
 
+use runkv_common::config::PrometheusConfig;
+use chrono::{DateTime, Duration, Local, NaiveDateTime};
+use http;
+use hyper::header::CONTENT_TYPE;
+use hyper::service::{make_service_fn, service_fn};
+use hyper::{Body, Server};
+use lazy_static::lazy_static;
+use log::info;
```
Comment (MrCroxx) [inline wheel/src/service.rs:46]: Ditto.

File: wheel/src/service.rs [hunk 20]
```diff
@@ -232,6 +245,24 @@ impl Wheel {
         Ok((group, 0, None))
     }
 }
+// prometheus
```
Comment (MrCroxx) [inline wheel/src/service.rs:248]: Remove the comment.

**Review by MrCroxx — CHANGES_REQUESTED** [review]

File: wheel/Cargo.toml [hunk 21]
```diff
@@ -10,13 +10,19 @@ async-trait = "0.1"
...
+http = "0.2.6"
 humantime = "2.1.0"
 humantime-serde = "1.1.1"
+hyper = { version = "^0.14", features = ["server", "http1", "tcp"] }
 itertools = "0.10.3"
+lazy_static = "^1.4"
 moka = { version = "0.7", features = ["future"] }
 openraft = { git = "https://github.com/datafuselabs/openraft", rev = "ca8a09c1898dbcaa4c2bf49bf5dabc5221e0b908" }
 parking_lot = "0.12"
+
```
Comment (MrCroxx) [inline wheel/Cargo.toml:24]: Remove the empty line. BTW, have you run `make fmt` first?

File: wheel/src/lib.rs [hunk 22]
```diff
@@ -53,6 +68,24 @@ pub async fn bootstrap_wheel(
         .map_err(Error::err)
 }
 
+pub fn boot_prometheus_service(listen_addr: String, wheel: Wheel) {
+    tokio::spawn(async move {
+        info!(
+            "Prometheus listener for Prometheus is set up on http://{}",
+            listen_addr
+        );
+        let listen_prometheus_addr: SocketAddr = listen_addr.parse().unwrap();
```
Comment (MrCroxx) [inline wheel/src/lib.rs:77]: Not sure if this type identifier can be removed.

File: Makefile.toml [hunk 23]
```diff
@@ -38,13 +38,15 @@ echo success
...
-dependencies = ["download-minio", "download-mcli", "clean-data"]
+dependencies = ["download-minio", "download-mcli", "download-prometheus", "download-grafana", "clean-data"]
 script = '''
 #!/bin/bash
 # run minio
 tmux new -d -s runkv-minio ${PREFIX_BIN}/minio server --address 127.0.0.1:9000 --console-address 127.0.0.1:9090 ${PREFIX_DATA}
 # run jaeger
 tmux new -d -s runkv-jaeger ${PREFIX_BIN}/jaeger
+# run prometheus
+tmux new -d -s runkv-prometheus "${PREFIX_BIN}/prometheus/prometheus" --config.file="${PWD}/tests/etc/prometheus.yml" --web.listen-address=:9091 --storage.tsdb.path="${PREFIX_DATA}/prometheus_data"
```
Comment (MrCroxx) [inline Makefile.toml:49]: `$PWD` should be removed. Or if we run `run` from other directory, it will refer to wrong directory. 🤔

**Review by MrCroxx — APPROVED** [review]
LGTM!

**Pull request merged**